include/glow/Optimizer/GraphOptimizer/NodeSplitting.h

```diff
@@ -0,0 +1,283 @@
+/**
+ * Copyright (c) Glow Contributors. See CONTRIBUTORS file.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_OPTIMIZER_GRAPHOPTIMIZER_NODESPLITTING_H
+#define GLOW_OPTIMIZER_GRAPHOPTIMIZER_NODESPLITTING_H
+
+#include "glow/Base/Type.h"
+#include "glow/Graph/Graph.h"
+#include "glow/Graph/Node.h"
+#include "glow/Support/Error.h"
+
+#include "llvm/ADT/ArrayRef.h"
+
+#include <functional>
+
+namespace glow {
+
+///===---------------------------------------------------------------------===//
+///                               SplitNodeOption
+///===---------------------------------------------------------------------===//
+/// Generic split node option as base type. All the split options refer to the
+/// split procedure performed with respect to the tensor associated with the
+/// output operand of a node. If a node has multiple output operands then the
+/// option commonly refers to the first output operand.
+class SplitNodeOption {
+
+  /// All the dimensions of a tensor used for splitting in increasing order.
+  std::vector<size_t> splitDims_;
```

Comment:
Could you use a `llvm::SmallVector<size_t, max_tensor_dimensions>` to avoid dynamic memory allocation?

Reply:
Question: The `max_tensor_dimensions` does limit the maximum number of elements or is just used for initially allocating the vector?

Reply:
It initially reserves memory for `most max_tensor_dimensions` on stack. If at runtime more elements are needed, it will grow and use heap allocation, just like a usual std::vector. You usually use SmallVector if you know you are going to have a small number of elements in your vector and you know the likely high watermark for this number in advance.

Reply:
Got it. I started making the replacements and I got a lot of errors. How do I initialize a SmallVector with an ArrayRef here:
```
SplitNodeOption(llvm::ArrayRef<size_t> splitDims) : splitDims_(splitDims) {}
```

Reply:
I guess, you should be able to use iterators: `splitDims_(splitDims.begin(), splitDims.end())`

Reply:
Done.